autogen/oai/client.py

```diff
@@ -442,21 +446,66 @@ def cost(self, response: Union[ChatCompletion, Completion]) -> float:
         return tmp_price1K * (n_input_tokens + n_output_tokens) / 1000
 
     @classmethod
-    def extract_text_or_function_call(cls, response: ChatCompletion | Completion) -> List[str]:
-        """Extract the text or function calls from a completion or chat response.
+    def extract_text_or_completion_object(
+        cls, response: ChatCompletion | Completion
+    ) -> Union[List[str], List[ChatCompletionMessage]]:
+        """Extract the text or ChatCompletion objects from a completion or chat response.
 
         Args:
             response (ChatCompletion | Completion): The response from openai.
 
         Returns:
-            A list of text or function calls in the responses.
+            A list of text, or a list of ChatCompletion objects if function_call/tool_calls are present.
         """
         choices = response.choices
         if isinstance(response, Completion):
             return [choice.text for choice in choices]
-        return [
-            choice.message if choice.message.function_call is not None else choice.message.content for choice in choices
-        ]
+
+        if TOOL_ENABLED:
+            return [
+                choice.message
+                if choice.message.function_call is not None or choice.message.tool_calls is not None
+                else choice.message.content
+                for choice in choices
+            ]
+        else:
+            return [
+                choice.message if choice.message.function_call is not None else choice.message.content
+                for choice in choices
+            ]
+
+    @staticmethod
+    def ChatCompletionMessage_to_dict(response: ChatCompletionMessage) -> Dict:
+        """Convert the response to a dict.
+
+        Args:
+            response (ChatCompletionMessagen): The ChatCompletionMessage from openai.
+
+        Returns:
+            A dict of the response.
+        """
+        if not isinstance(response, ChatCompletionMessage):
+            return response
+
+        dict_msg = {
+            "content": response.content,
+            "role": response.role,
+        }
+        if response.function_call is not None:
+            dict_msg["function_call"] = dict(response.function_call)
+
+        # TODO: in the future openai may support tool call to Code Interpreter, doc retriever, etc.
+        if TOOL_ENABLED and response.tool_calls is not None:
+            dict_msg["tool_calls"] = [
+                {
+                    "type": tool_call.type,
+                    "id": tool_call.id,
+                    "function": dict(tool_call.function),
+                }
+                for tool_call in response.tool_calls
+            ]
```

Comment:
I think there is a simpler method like to_dict_recursive()

Reply:
```
result = openai.ChatCompletion.create(
    messages=messages, model="gpt-3.5-turbo", temperature=0
)
result["choices"][0]["message"].to_dict_recursive()
```
Find some references: https://python.langchain.com/docs/integrations/adapters/openai-old
It is supported in the old one. I also checked the new response, it doesn't offer this feature.

Reply:
Changing to `.model_dump(mode="dict")`